### ISSUE TITLE
Retaining viewport item positions in virtual scroll

### ISSUE THREAD

**Issue by ariel-compie** [angular/components#12002]
#### Bug, feature request, or proposal:
Retaining viewport item positions in virtual scroll

#### What is the expected behavior?
Retaining viewport item positions in virtual scroll with "autosize" property after adding items to the beginning of the list

#### What is the current behavior?
After adding items to the beginning of the list the viewport jumps to a new position (when items are on autosize).

#### What are the steps to reproduce?
Providing a StackBlitz reproduction is the *best* way to share your issue. <br/>

https://stackblitz.com/edit/angular-cdk-demo-virtual-scroll-xxywis?file=app%2Fvirtual-scroll%2Fvirtual-scroll.component.ts

#### What is the use-case or motivation for changing an existing behavior?
Supporting chat like applications like Whatsapp - Lazy loading messages from server

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
latest cdk-experimental

**Comment by agarzas** [comment]
It also happens when the items are on fixed size

**Comment by agarzas** [comment]
Hi @mmalerba, I´m wondering if this functionality has been added to recent releases, is this going to be implemented?

**Comment by mmalerba** [comment]
I'm not sure if this is something that should be handled by the virtual scroll itself. It seems to me like some logic that would be specific to an app (e.g. scroll to bottom whenever the item list changes).

**Comment by agarzas** [comment]
Imagine that we are building a WhatsApp clone, and we open a long chat history that has 3007 Messages.

![chat_last_messages](https://user-images.githubusercontent.com/43526819/77720476-a8308980-6fad-11ea-8a0a-5c774a33a8e7.jpg)


The last message that I received was “Message 3007”.

To keep things simple the object containing all messages is an array of string, and when we load our component our array look like this:

```
let messages = [
    “Message 3007”,
    “Message 3006”,
    “Message 3005”,
    “Message 3004”,
    “Message 3003”,
    “Message 3002”,
];
```

Note that position [0] is the last message that I received, it is designed that way because we know that if we scroll up, we can append previously sent meesages that we store in some kind of database and the cdk-virtual-scroll component will handle the rendering beautifuly and efficiently.

Now, imagine that we scroll up and load a lot of meesages.

![chat_previously_messages](https://user-images.githubusercontent.com/43526819/77720530-d0b88380-6fad-11ea-910d-7835aff2a49e.jpg)


Now the array object looks like this:

```
let messages = [
    “Message 3007”,
    “Message 3006”,
    “Message 3005”,
    “Message 3004”,
    “Message 3003”,
    “Message 3002”,
    “…………………",
    “……………..…",
    “..………………",
    “Message 23”,
    “Message 22”,
    “Message 21”,
    “Message 20”,
    “Message 19”,
    “Message 18”,
    “Message 17”,
    “Message 16”,
    “Message 15”,
    “Message 14”,
];

```
What would happen if through websockets we received a new message “Message 3008” while we are reading “Message 14"?,
well we would have to append the new message at the begining of the array at position 0 and our screen cant retain the viewport position because we modified the begining of the array.

In other words, virtual scroll is only one-direction virtualized, and it would be nice if we could have a bidirectional scroller.

**Comment by mmalerba** [comment]
I see, I think what it comes down to is that there's a few different styles of virtual scrolling. The one we have implemented is intended to behave as closely as possible to a normal scrolling list (i.e. div with overflow:auto), but under the hood it doesn't render all of the elements. If we consider what would happen when adding an element at the top in a normal scrolling div, we would expect it to push the content down.

This chat style you're describing may be more similar to an infinite scrolling list. Though "virtual" and "infinite" are often used interchangeably, they are subtly different. I would suggest maybe looking into some third party solutions for infinite scrolling and see if any of those meet your needs.

**Comment by mmalerba** [comment]
Closing because this feature doesn't make sense for virtual scroll. If we decide to do an infinite scroll at some point in the future we will add support for this
